Document preview PDF generation events for Emails and Landing Pages (#508)
mautic/developer-documentation-new: docs/plugin_extensions/emails.rst

This section is in progress. See  ``\Mautic\EmailBundle\Stats\Helper\StatHelperInterface``

> Preview PDF generation
> ----------------------

Reviewer: ```suggestion
.. vale off

Preview PDF generation
----------------------

.. vale on
```

> Users can download Email previews as PDF files. Mautic uses an event-driven PDF generation process, enabling plugins and distributions to provide custom PDF renderers.

Reviewer: There are a couple of things here to address the Vale warnings:

- Capitalize "Plugins"
- Find alternative word for "renderers" that suitable for the context

> Mautic dispatches the ``\Mautic\EmailBundle\EmailEvents::EMAIL_PREVIEW_GENERATE_PDF`` event when a user downloads an Email preview as a PDF. The default community implementation uses Dompdf when available, falling back to a basic stdlib-based PDF generator.

Reviewer: ```suggestion
Mautic dispatches the ``\Mautic\EmailBundle\EmailEvents::EMAIL_PREVIEW_GENERATE_PDF`` event when a User downloads an Email preview as a PDF. The default community implementation uses Dompdf when available, falling back to a basic stdlib-based PDF generator.
```

Event details
=============

* **Event name:** ``mautic.email_preview_generate_pdf``
* **Constant:** ``\Mautic\EmailBundle\EmailEvents::EMAIL_PREVIEW_GENERATE_PDF``
* **Event class:** ``\Mautic\EmailBundle\Event\EmailPreviewPdfGenerationEvent``

Default implementation
======================

Mautic provides default PDF generation through:

* **Subscriber:** ``\Mautic\EmailBundle\EventListener\EmailPreviewPdfSubscriber``
* **Generator:** ``\Mautic\EmailBundle\Pdf\EmailPreviewPdfGenerator``

The default subscriber runs at priority ``-255``, allowing custom implementations to take precedence.

If no listener sets PDF content, the controller returns a 500 error.

Routes
======

Email preview PDF downloads use these routes:

* ``/email/preview/{objectId}/download/{downloadType}``
* ``/email/preview/{objectId}/draft/download/{downloadType}``

The ``downloadType`` parameter defaults to ``pdf``.

Event payload
=============

The event provides access to:

* ``getHtmlContent()`` - The rendered HTML preview content
* ``getEmail()`` - The Email entity
* ``getContact()`` - The Contact context, if selected for preview

> * ``getRequest()`` - The HTTP Request object

Reviewer: ```suggestion
* ``getRequest()`` - The HTTP request object
```

* ``getFileName()`` - The suggested PDF filename

Event methods
=============

To provide a custom PDF:

1. Generate binary PDF bytes from the HTML content using ``getHtmlContent()``
2. Call ``setPdfContent($bytes)`` to set the PDF content
3. Optionally call ``setFileName()`` to customize the output filename

> 4. Use ``hasPdfContent()`` to check if another subscriber already provided PDF content

Reviewer: There are a couple of things here:

- Follow Mautic style guide for ordered list by using `#`.
- Find another word of "check"

```suggestion
#. Generate binary PDF bytes from the HTML content using ``getHtmlContent()``
#. Call ``setPdfContent($bytes)`` to set the PDF content
#. Optionally call ``setFileName()`` to customize the output filename
#. Use ``hasPdfContent()`` to check if another subscriber already provided PDF content
```

> Custom PDF generator example
> ============================

Reviewer: ```suggestion
.. vale off

Custom PDF generator example
============================

.. vale on
```

This section is in progress. See  ``\Mautic\EmailBundle\Stats\Helper\StatHelperInterface``

Preview PDF generation

> ----------------------

Author: Event details, class names (`EmailPreviewPdfGenerationEvent`, `EmailEvents::EMAIL_PREVIEW_GENERATE_PDF`), and default subscriber priority (-255) from the PR's new `EmailBundle/Event/EmailPreviewPdfGenerationEvent.php` and `EmailBundle/EventListener/EmailPreviewPdfSubscriber.php` files.

Source: https://github.com/mautic/mautic/pull/16187

Event details
=============

* **Event name:** ``mautic.email_preview_generate_pdf``
* **Constant:** ``\Mautic\EmailBundle\EmailEvents::EMAIL_PREVIEW_GENERATE_PDF``
* **Event class:** ``\Mautic\EmailBundle\Event\EmailPreviewPdfGenerationEvent``

Default implementation
======================

Mautic provides default PDF generation through:

* **Subscriber:** ``\Mautic\EmailBundle\EventListener\EmailPreviewPdfSubscriber``
* **Generator:** ``\Mautic\EmailBundle\Pdf\EmailPreviewPdfGenerator``

The default subscriber runs at priority ``-255``, allowing custom implementations to take precedence.

If no listener sets PDF content, the controller returns a 500 error.

Routes
======

Email preview PDF downloads use these routes:

* ``/email/preview/{objectId}/download/{downloadType}``
* ``/email/preview/{objectId}/draft/download/{downloadType}``

The ``downloadType`` parameter defaults to ``pdf``.

Event payload
=============

The event provides access to:

> * ``getHtmlContent()`` - The rendered HTML preview content
> * ``getEmail()`` - The Email entity
> * ``getContact()`` - The Contact context, if selected for preview
> * ``getRequest()`` - The HTTP request object
> * ``getFileName()`` - The suggested PDF filename

Reviewer: ```suggestion
.. vale off

* ``getHtmlContent()`` - The rendered HTML preview content
* ``getEmail()`` - The Email entity
* ``getContact()`` - The Contact context, if selected for preview
* ``getRequest()`` - The HTTP request object
* ``getFileName()`` - The suggested PDF filename

.. vale on
```
